libs/ts-rest/core/src/lib/client.ts

```diff
@@ -99,16 +100,31 @@ export const tsRestFetchApi: ApiFetcher = async ({
     credentials,
     signal,
     cache,
-    next
+    next,
     // we must type cast here because the typescript types for RequestInit
     // do not include properties like "next" for Frameworks (like Nextjs)
   } as RequestInit);
   const contentType = result.headers.get('content-type');
 
-  if (contentType?.includes("application/") && contentType?.includes('json')) {
+  if (contentType?.includes('application/') && contentType?.includes('json')) {
+    if (!route.validateResponseOnClient) {
+      return {
+        status: result.status,
+        body: await result.json(),
+        headers: result.headers,
+      };
+    }
+
+    const jsonData = await result.json();
+    const statusCode = result.status;
+    const response = route.responses[statusCode];
+
     return {
-      status: result.status,
-      body: await result.json(),
+      status: statusCode,
+      body:
+        response && typeof response !== 'symbol' && 'parse' in response
+          ? response?.parse(jsonData)
```

Comment:
are we sure we want to do `parse` instead of `safeParse`? I feel like we should allow the client consumers to have more flexibility in their error handling? Or maybe it makes sense to just throw and let the client catch the error?

Reply:
`safeParse` could be used here, but that presents a couple problems.
1. What happens if it doesn't pass validation? What happens to the flow then?
2. How do you convey the details of the error to the client? (For those that either want to decorate their forms or show a toast/banner)

Unless the plan is to offer some kind of `clientOnError` function (which probably wouldn't be feasible since the contract also needs to be compatible for sharing on the server, and the contract API surface would be broadening), you'd then probably have to throw a custom TsRest flavoured Error.
If you do go down the custom Error thrown approach, you'd need to have properties on it for returning the original error data and to distinguish the type of validator being used (i.e: zod, internal, custom, ....).

In its current implementation, you are able to check if the error thrown is an `instanceof ZodError` and deal with it accordingly.

That was my rationale for using the 'parse' fn with zod.

Reply:
I completely agree, just wanted to make sure we had this discourse. 

Can you add some documentation to this feature in this PR or a subsequent PR and mention that it will throw (and you can check for `instanceof ZodError`)?

Reply:
Sure thing!